src/panels/TimePanel/TimeInput/TimeInput.tsx

```diff
@@ -0,0 +1,267 @@
+    setDate(luaApi, pendingTime);
+  }
+
+  function setTime(data: {
+    time: Date;
+    immediate: boolean; // If not, interpolate
+    delta: number;
+    relative: boolean;
+  }) {
+    if (!isDateValid(data.time)) {
+      return;
+    }
+
+    if (useLock) {
+      // If we're in locked mode, just set the pending timestamp. this timestamp will
+      // be used when the user clicks the "interpolate" or "set" button
+      setPendingTime(data.time);
+      return;
+    }
+
+    if (data.immediate) {
+      setDate(luaApi, data.time);
+      return;
+    }
+
+    // Interpolate
+    if (data.relative) {
+      luaApi?.time.interpolateTimeRelative(data.delta);
+    } else {
+      interpolateDate(data.time);
+    }
+  }
+
+  function onTimeInputRelative(timePart: TimePart, change: number) {
+    // Update the current time stamp with the change
+    const newTime = new Date(timeRef.current);
+    switch (timePart) {
+      case TimePart.Milliseconds:
+        newTime.setUTCMilliseconds(newTime.getUTCMilliseconds() + change);
+        break;
+      case TimePart.Seconds:
+        newTime.setUTCSeconds(newTime.getUTCSeconds() + change);
+        break;
+      case TimePart.Minutes:
+        newTime.setUTCMinutes(newTime.getUTCMinutes() + change);
+        break;
+      case TimePart.Hours:
+        newTime.setUTCHours(newTime.getUTCHours() + change);
+        break;
+      case TimePart.Days:
+        newTime.setUTCDate(newTime.getUTCDate() + change);
+        break;
+      case TimePart.Months:
+        newTime.setUTCMonth(newTime.getUTCMonth() + change);
+        break;
+      case TimePart.Years:
+        newTime.setUTCFullYear(newTime.getUTCFullYear() + change);
+        break;
+      default:
+        throw Error(`Unhandled 'TimePart' case: ${timePart}`);
+    }
+
+    setTime({
+      time: newTime,
+      immediate: isHoldingShift,
+      delta: (newTime.getTime() - timeRef.current.getTime()) / 1000,
+      relative: true
+    });
+  }
+
+  function onTimeInput(timePart: TimePart, value: number) {
+    const newTime = new Date(timeRef.current);
+    switch (timePart) {
+      case TimePart.Milliseconds:
+        newTime.setUTCMilliseconds(value);
+        break;
+      case TimePart.Seconds:
+        newTime.setUTCSeconds(value);
+        break;
+      case TimePart.Minutes:
+        newTime.setUTCMinutes(value);
+        break;
+      case TimePart.Hours:
+        newTime.setUTCHours(value);
+        break;
+      case TimePart.Days:
+        newTime.setUTCDate(value);
+        break;
+      case TimePart.Months:
+        newTime.setUTCMonth(value);
+        break;
+      case TimePart.Years:
+        newTime.setUTCFullYear(value);
+        break;
+      default:
+        throw Error(`Unhandled 'TimePart' case: ${timePart}`);
+    }
+
+    setTime({
+      time: newTime,
+      immediate: true,
+      delta: 0, // This variable is not relevant in this case
+      relative: false
+    });
+  }
+
+  return cappedTime ? (
```

Comment:
Is this checking for undefined? Maybe make this a bit clearer, like an early out of the function -  I didn't realize until I got to the end of the file that it was a check for validity

Reply:
seeing this now, I agree we should move it to an early out